src/sources/postgresql_metrics.rs

```diff
@@ -0,0 +1,851 @@
+        match self {
+            Self::Include {
+                need_null,
+                databases,
+            } => {
+                conditions += " WHERE datname = ANY($1)";
+                if *need_null {
+                    conditions += " OR datname IS NULL";
+                }
+                client
+                    .query(conditions.as_str(), &[&databases.as_slice()])
+                    .await
+            }
+            Self::Exclude {
+                need_null,
+                databases,
+            } => {
+                conditions += " WHERE datname != ANY($1)";
+                if *need_null {
+                    conditions += " AND datname IS NOT NULL";
+                } else {
+                    conditions += " OR datname IS NULL";
+                }
+                client
+                    .query(conditions.as_str(), &[&databases.as_slice()])
+                    .await
+            }
+            Self::None => client.query(conditions.as_str(), &[]).await,
+        }
+    }
+
+    async fn pg_stat_database_conflicts(&self, client: &Client) -> Result<Vec<Row>, PgError> {
+        let mut conditions = "SELECT * FROM pg_stat_database_conflicts".to_owned();
+        match self {
+            Self::Include { databases, .. } => {
+                conditions += " WHERE datname = ANY($1)";
+                client
+                    .query(conditions.as_str(), &[&databases.as_slice()])
+                    .await
+            }
+            Self::Exclude { databases, .. } => {
+                conditions += " WHERE datname != ANY($1)";
+                client
+                    .query(conditions.as_str(), &[&databases.as_slice()])
+                    .await
+            }
+            Self::None => client.query(conditions.as_str(), &[]).await,
+        }
+    }
+}
+
+#[derive(Debug)]
+struct PostgresqlMetrics {
+    config: Config,
+    tls_config: Option<PostgresqlMetricsTlsConfig>,
+    client: Option<Client>,
+    namespace: Option<String>,
+    tags: BTreeMap<String, String>,
+    datname_filter: DatnameFilter,
+}
+
+impl PostgresqlMetrics {
+    async fn new(
+        endpoint: String,
+        datname_filter: DatnameFilter,
+        namespace: Option<String>,
+        tls_config: Option<PostgresqlMetricsTlsConfig>,
+    ) -> Result<Self, BuildError> {
+        let config: Config = endpoint.parse().context(InvalidEndpoint)?;
+        if config.get_hosts().is_empty() {
+            return Err(BuildError::HostMissed);
+        }
+        if config.get_hosts().len() > 1 {
+            return Err(BuildError::MultipleHostsNotSupported {
+                hosts: config.get_hosts().to_owned(),
+            });
+        }
+
+        let mut tags = BTreeMap::new();
+        tags.insert("endpoint".into(), config_to_endpoint(&config));
+        tags.insert(
+            "host".into(),
+            match &config.get_hosts().get(0).expect("host is not missed") {
```

Comment:
This isn't bad since we check this invariant earlier, but I might just do something like:

```
&config.get_hosts().get(0).unwrap_or("unknown")
```

To avoid panic'ing.

Alternatively, we could store the config differently such that, when the invariant of 1 host is asserted, we pull it out and put it alongside the config so that other places can depend on it existing.